probes/pkg/dns/probe.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dns
+
+import (
+	"flag"
+	"net"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/klog"
+)
+
+const (
+	namespace = "probes"
+)
+
+var (
+	// inClusterDNSLatency implements the In-Cluster DNS latency SLI, see
+	// https://github.com/kubernetes/community/blob/master/sig-scalability/slos/dns_latency.md
+	inClusterDNSLatency = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: namespace,
+		Name:      "in_cluster_dns_latency_seconds",
+		Buckets:   prometheus.ExponentialBuckets(0.000001, 2, 26), // from 1us up to ~1min
+		Help:      "Histogram of the time (in seconds) it took to ping a ping-server instance.",
+	})
+	// inClusterDNSLookupCount counts DNS lookups done by prober.
+	inClusterDNSLookupCount = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: namespace,
+		Name:      "in_cluster_dns_lookup_count",
+		Help:      "Counter of DNS lookups made by dns-prober.",
+	})
+	// inClusterDNSLookupError counts failed DNS lookups done by prober.
+	inClusterDNSLookupError = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: namespace,
+		Name:      "in_cluster_dns_lookup_error",
+		Help:      "Counter of DNS lookups made by dns-prober that failed",
+	})
+
+	probe dnsProbe
+)
+
+func init() {
+	prometheus.MustRegister(inClusterDNSLatency, inClusterDNSLookupCount, inClusterDNSLookupError)
+
+	flag.StringVar(&probe.lookupURL, "dns-probe-url", "", "Name of a Service to lookup")
+	flag.DurationVar(&probe.probingInterval, "probing-interval", 1*time.Second, "Interval between DNS lookups")
+}
+
+type dnsProbe struct {
+	lookupURL       string
+	probingInterval time.Duration
+}
+
+// Run periodically does DNS lookups.
+// Lookup's interval and URL are configurable via flags.
+func Run() {
+	if probe.lookupURL == "" {
+		klog.Fatal("--dns-probe-url has not been set")
+	}
+	probe.run()
+}
+
+func (p *dnsProbe) run() {
```

Comment:
You may want to pass stopCh to it.

Reply:
This is not used by other probes. I'd rather add it in another PR.